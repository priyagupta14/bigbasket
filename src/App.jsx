/* eslint-disable no-unused-vars */
import './App.css';
import React, { useEffect, useState } from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import axios from 'axios';
import banana from './assets/images/banana.jpg';
import setting from './assets/images/setting.png';
import Home from './Components/Home/Home';
import NavBar from './Components/NavBar/NavBar';
import Cart from './Components/Cart/Cart';
import AllOrders from './Components/AllOrders/AllOrders';
import Checkout from './Components/Checkout/Checkout';
import { ThemeContext } from './ThemeContext';
// import HookCounter from './Components/HookCounter/HookCounter';

export default function App() {
  const [AllProduct, setProduct] = useState([]);
  const [cartCount, setCartCount] = useState(0);
  const [cartItem, setCartItem] = useState([]);
  const [theme, setTheme] = useState('light');
  const [categorized, setCategories] = useState({});

  const [error1, setError] = useState(null);
  const [isLoaded, setIsLoaded] = useState(false);
  const [pastOrder, setPastOrder] = useState([
    {
      orderId: 1,
      noOfItems: 3,
      date: 'Sun 04 Mar 2018',
      amount: 120,
      items: [
        {
          id: 1,
          name: 'Banana',
          count: 0,
          quantity: 1,
          price: 40,
        },
        {
          id: 2,
          image: './assets/apple.png',
          name: 'Apple',
          count: 0,
          quantity: 1,
          price: 40,
        },
      ],
    },
    {
      orderId: 2,
      noOfItems: 2,
      date: 'Sun 03 Mar 2018',
      amount: 80,
      items: [
        {
          id: 1,
          image: './assets/banana.png',
          name: 'Banana',
          count: 0,
          quantity: 1,
          price: 40,
        },
      ],
    },
  ]);

  const groupByCategory = (items) => items.reduce((acc, product) => {
    const { category } = product;
    if (!acc[category]) {
      acc[category] = [];
    }
    acc[category].push(product);
    return acc;
  }, {});

  useEffect(async () => {
    const { data, error } = await axios.get('/items');
    const products = data.data;
    if (products) setIsLoaded(true);
    else { setError(error); setIsLoaded(true); }
    const allProducts = products.map((eachProduct) => ({
      ...eachProduct,
      inCartCount: 0,
      url: banana,
    }));
    const categorizedProduct = groupByCategory(allProducts);
    setCategories(categorizedProduct);
    setProduct(allProducts);
  }, []);

  const onDecrement = (id, category) => {
    const newProduct = categorized[category].map((product) => (product.id === id ? {
      ...product,
      inCartCount: product.inCartCount > 0 ? product.inCartCount - 1 : product.inCartCount,
      count: product.inCartCount > 0 ? product.count + 1 : product.count,
    } : product));
    categorized[category] = newProduct;
    setCategories(categorized);
    const newCount = (categorized[category].find(
      (product) => { if (product.id === id) return product.inCartCount > 0; return false; },
    )) ? cartCount - 1 : cartCount;
    setCartCount(newCount);
    const newCartItem = categorized[category].filter((product) => product.inCartCount > 0);
    setCartItem(newCartItem);
  };

  const onIncrement = (id, category) => {
    const newProduct = categorized[category].map((eachProduct) => (
      eachProduct.id === id ? {
        ...eachProduct,
        inCartCount: eachProduct.count > 0 ? eachProduct.inCartCount + 1 : eachProduct.inCartCount,
        count: eachProduct.count > 0 ? eachProduct.count - 1 : eachProduct.count,
      } : eachProduct
    ));
    categorized[category] = newProduct;
    setCategories(categorized);

    const newCount = (categorized[category].find(
      (product) => {
        if (product.id === id) return (product.count > 0 && product.inCartCount > 0); return false;
      },
    )) ? cartCount + 1 : cartCount;
    setCartCount(newCount);
    const newCartItem = categorized[category].filter((product) => product.inCartCount > 0);
    setCartItem(newCartItem);
  };

  if (error1) {
    return (
      <div>error1.message</div>
    );
  }
  if (!isLoaded) {
    return (
      <div>Loading...</div>
    );
  }
  return (
    <div>
      <BrowserRouter>
        <ThemeContext.Provider value={theme}>
          <NavBar itemInCart={cartCount} />
        </ThemeContext.Provider>
        <button type="button" className="toggle-button" onClick={() => setTheme(theme === 'light' ? 'dark' : 'light')}>
          <img src={setting} alt="setting" />
        </button>
        <Switch>
          <Route path="/" exact>
            <Home
              categorizedProduct={categorized}
              productList={AllProduct}
              onIncrement={onIncrement}
              onDecrement={onDecrement}
            />
          </Route>
          <Route path="/cart">
            <Cart productList={cartItem} />
          </Route>
          <Route path="/allOrders">
            <AllOrders order={pastOrder} />
          </Route>
          <Route path="/checkout">
            <Checkout />
          </Route>
        </Switch>
      </BrowserRouter>
    </div>
  );
}
