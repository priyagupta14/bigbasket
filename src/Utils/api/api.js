import axios from 'axios';

export const getItems = async () => {
  try {
    const allItems = await axios.get('/items');
    return allItems.data.data;
  } catch (error) {
    return error;
  }
};
export const getOrders = async () => {
  try {
    const allOrders = await axios.get('/orders');
    return allOrders.data.data;
  } catch (error) {
    return error;
  }
};
export const postOrders = async (orders) => {
  try {
    console.log(orders);
    const allOrders = await axios.post('/orders', orders);
    return allOrders.data;
  } catch (error) {
    return error;
  }
};
// export default { getItems, getOrders };
