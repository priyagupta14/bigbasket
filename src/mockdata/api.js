/* eslint-disable import/prefer-default-export */
export const mockProducts = [{
  id: 1,
  name: 'apple',
  price: 120,
  count: 20,
  category: 'Fruits & Vegatables',
}];

export const mockOrders = {
  items: [
    {
      id: 1,
      name: 'apple',
      price: 120,
      count: 1,
      category: 'Fruits & Vegatables',
    },
  ],
  id: 1,
  date: 1615122360481,
};
export const mockPostOrder = {
  items: [
    {
      id: 4,
      name: 'duster',
      price: 80,
      count: 1,
      category: 'Household Items',
    },
    {
      id: 5,
      name: 'milk',
      price: 10,
      count: 2,
      category: 'Dairy & Eggs',
    },
  ],
  id: 25,
  date: 1615526811454,
};
export const mockOrderedOrder = {
  items: [
    {
      id: 2, name: 'pp', count: 1, category: 'idk', price: 24,
    },
    {
      id: 3, name: 'bp', count: 1, category: 'idk123', price: 25,
    },
  ],
};
