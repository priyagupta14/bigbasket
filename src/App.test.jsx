import {
  fireEvent, render, screen, waitFor,
} from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import App from './App';
import { mockOrderedOrder, mockPostOrder, mockProducts } from './mockdata/api';
import axiosApi from './Utils/api/api';

describe(App.name, () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    jest.spyOn(axiosApi, 'getItems').mockResolvedValue(mockProducts);
  });

  // let mockGetItems= jest.spyOn();
  // beforeEach(() => {
  //   getItems.mockResolvedValue(mockProducts);
  //   getItems.mockClear();
  // });
  // test('should match snapShot', async () => {
  //   const { container } = await waitFor(() => render(<App />));
  //   expect(container).toMatchSnapshot();
  //   // await waitFor(() => render(<App />));
  // });
  test('should match the snapshot', async () => {
    const { container } = await waitFor(() => render(
      <BrowserRouter>
        <App />
      </BrowserRouter>,
    ));
    expect(container).toMatchSnapshot();
  });
  xtest('should render Home component and route t', () => {

  });
});
