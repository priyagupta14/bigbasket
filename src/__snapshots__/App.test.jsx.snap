// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div>
  <div>
    <div
      class="light-navbar"
    >
      <a
        href="/"
        style="text-decoration: none; color: white;"
      >
        <br />
        E-Shopper
      </a>
      <div
        class="left-side-navbar"
      >
        <a
          href="/allOrders"
          style="text-decoration: none; color: white;"
        >
          <br />
          All Orders
        </a>
        <a
          href="/cart"
          style="text-decoration: none; color: white;"
        >
          <p>
            MyBasket
          </p>
          <div
            class="myBasket-container"
          >
            <p>
              MyCart
              0
            </p>
          </div>
        </a>
      </div>
    </div>
    <button
      class="toggle-button"
      type="button"
    >
      <img
        alt="setting"
        src="setting.png"
      />
    </button>
    <div>
      <div>
        <h1>
          Fruits & Vegatables
        </h1>
        <div
          class="product-card"
        >
          <div
            class="product-card-element"
          >
            <img
              alt="fruitItem"
              data-testid="product-img"
              src="banana.jpg"
            />
            <p>
              apple
            </p>
            <div
              class="price-counter"
            >
              <p>
                MRP
                120
                /-
              </p>
              <div
                class="zeroProduct"
              >
                <button
                  type="button"
                >
                          +      
                </button>
                <p>
                  0
                   
                  in basket
                </p>
                <button
                  type="button"
                >
                  -
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
